### DISCUSSION TITLE
Why Linq2db Table class is sealed? Or how implement base filtering for generic tables

### DISCUSSION THREAD

**Discussion by rovdima** [linq2db/linq2db#4150] [Q&A] [answered]
Hello everyone, 
hope you are doing well!

I need in [TableT](https://github.com/linq2db/linq2db/blob/738fd905cb187763a72c4d0d64b5c427772b8ab5/Source/LinqToDB/Linq/TableT.cs)
for build own base Table-repository which can have some base 'where'-filtering for my ITable<T>.

So, previously, using BlToolkit, we used Table<T> and inherited from it... As example.
For now TableT is sealed.

We decided to port our Old Legacy Project to LinqToDb and want to implement the same behaviour
How it possible now?

So, for example,

I generated ITables<T> (ex. ITable for SomeEntity) from template as recommended
I use Linq for some queries like this
db.SomeEntities.Where(x=>x.Id = 1)
But I want to implement some base repository and get following
db.SomeEntities.GetActualItems().Where(x=>x.Id = 1),
where GetActualItems() - some 'where'-filtering (ex .Where(x=>x.ActualVersion = 'v2'))

Do you have any ideas how it possible to do?

Thank you very much for any ideas.

**Comment by jods4** [comment, marked as answer]
I usually do it by adding top-level functions in my DB connection/context directly.
Instead of `db.SomeEntities.GetActualItems().Where(x => x.Id == 1)` I would have `db.GetActualItems().Where(x => x.Id == 1)`.

Extensions methods, as well as generic methods, are other options, esp. if you want to filter multiple collections with the same code.
For example if you have a soft-delete flag you could do:
```csharp
// Extension method:
public static IQueryable<T> DeletedItems<T>(this IQueryable<T> src) where T: ISoftDelete
  => src.Where(x => x.IsDeleted);
  
// Then you can do:
db.SomeStuff.DeletedItems().First(x => x.Id == 42);

// You could do the same with a generic method on DataConnection/DataContext though:
public IQueryable<T>  Deleted<T>() where T: ISoftDelete
  => this.GetTable<T>().Where(x => x.IsDeleted);
  
// Use like this:
db.Deleted<SomeStuff>().First(x => x.Id == 42);
```

Unless you want to change the built-in `ITable` behavior, you don't need to derive from it.
